### PR TITLE
chore(rejection tags): Session rejection tags

### PR DESCRIPTION
**Pull request by jackpooleywc**
# Description

This PR adds specific rejection tags for `sessionPropose` and `sessionAuthenticate` to disambiguate user rejection from genuine failures. 

## How Has This Been Tested?

This will be manually tested.

### CODE REVIEW

File: Sources/WalletConnectSign/Types/ProtocolMethods/SessionProposeProtocolMethod.swift [hunk 1]
```diff
@@ -2,8 +2,45 @@ import Foundation
 
 struct SessionProposeProtocolMethod: ProtocolMethod {
     let method: String = "wc_sessionPropose"
+    
+    enum Tag: Int {
+        case sessionPropose = 1100
+        case sessionProposeResponseApprove = 1101
+        case sessionProposeResponseReject = 1120
+        case sessionProposeResponseAutoReject = 1121
```
Comment (llbartekll) [inline Sources/WalletConnectSign/Types/ProtocolMethods/SessionProposeProtocolMethod.swift:10]: 👍

File: Sources/WalletConnectSign/Types/ProtocolMethods/SessionProposeProtocolMethod.swift [hunk 2]
```diff
@@ -2,8 +2,45 @@ import Foundation
...
+            prompt: false,
+            ttl: Int(ttl)
+        )
+    }
+    
+    static func responseApprove(ttl: TimeInterval = Self.defaultTtl) -> Self {
+        Self(ttl: ttl, responseTag: .sessionProposeResponseApprove)
+    }
+    
+    static func responseReject(ttl: TimeInterval = Self.defaultTtl) -> Self {
```
Comment (llbartekll) [inline Sources/WalletConnectSign/Types/ProtocolMethods/SessionProposeProtocolMethod.swift:39]:
```suggestion
    static func responseUserReject(ttl: TimeInterval = Self.defaultTtl) -> Self {
```
would it be better to be more specific?

File: Sources/WalletConnectSign/Auth/Services/App/SessionAuthRequestService.swift [hunk 3]
```diff
@@ -29,7 +29,7 @@ actor SessionAuthRequestService {
         var params = params
         let pubKey = try kms.createX25519KeyPair()
         let responseTopic = pubKey.rawRepresentation.sha256().toHexString()
-        let protocolMethod = SessionAuthenticatedProtocolMethod(ttl: params.ttl)
+        let protocolMethod = SessionAuthenticatedProtocolMethod.responseApprove(ttl: params.ttl)
```
Comment (llbartekll) [inline Sources/WalletConnectSign/Auth/Services/App/SessionAuthRequestService.swift:32]: here we have a request not response, should not have an effect on the behaviour but this init reads strangely.

File: Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift [hunk 4]
```diff
@@ -36,7 +36,7 @@ class AuthRequestSubscriber {
     }
 
     private func subscribeForRequest() {
-        pairingRegisterer.register(method: SessionAuthenticatedProtocolMethod())
+        pairingRegisterer.register(method: SessionAuthenticatedProtocolMethod.responseApprove())
```
Comment (llbartekll) [inline Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift:39]: same here

File: Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift [hunk 5]
```diff
@@ -73,7 +73,12 @@ class AuthRequestSubscriber {
     func respondError(payload: SubscriptionPayload, reason: SignReasonCode) {
         Task(priority: .high) {
             do {
-                try await networkingInteractor.respondError(topic: payload.topic, requestId: payload.id, protocolMethod: SessionAuthenticatedProtocolMethod(), reason: reason)
+                try await networkingInteractor.respondError(
```
Comment (llbartekll) [inline Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift:76]: let's mark this func private
```    
private func respondError(payload: SubscriptionPayload, reason: SignReasonCode) {
```
to be sure that's it's not called from outside with this tag

File: Sources/WalletConnectSign/Engine/Common/ApproveEngine.swift [hunk 6]
```diff
@@ -215,12 +220,14 @@ final class ApproveEngine {
 private extension ApproveEngine {
 
     func setupRequestSubscriptions() {
-        pairingRegisterer.register(method: SessionProposeProtocolMethod())
+        pairingRegisterer.register(method: SessionProposeProtocolMethod.responseAutoReject())
             .sink { [unowned self] (payload: RequestSubscriptionPayload<SessionType.ProposeParams>) in
                 guard let pairing = pairingStore.getPairing(forTopic: payload.topic) else { return }
+                let responseApproveMethod = SessionAuthenticatedProtocolMethod.responseApprove().method
```
Comment (llbartekll) [inline Sources/WalletConnectSign/Engine/Common/ApproveEngine.swift:226]: also feels strange that u need to init with responseApprove() to just get a method string

**Review by llbartekll — COMMENTED** [review]
left a few comments but the logic looks correct 👍

File: Sources/WalletConnectSign/Types/ProtocolMethods/SessionProposeProtocolMethod.swift (hunk 2 above)
Comment (jackpooleywc) [reply Sources/WalletConnectSign/Types/ProtocolMethods/SessionProposeProtocolMethod.swift:39]: I'm using `responseReject` for user rejected and `responseAutoReject` for protocol level auto rejections. I agree that this naming could be clearer but I also wanted to keep things aligned with [`Kotlin`](https://github.com/WalletConnect/WalletConnectKotlinV2/pull/1394/files#diff-c384e6db5880e14f35640a2d0a5ca3929714398868a15de74ce79c87cc8d4c85R14)

File: Sources/WalletConnectSign/Auth/Services/App/SessionAuthRequestService.swift (hunk 3 above)
Comment (jackpooleywc) [reply Sources/WalletConnectSign/Auth/Services/App/SessionAuthRequestService.swift:32]: I agree, and I started out by trying to change [`method`](https://github.com/WalletConnect/WalletConnectSwiftV2/blob/3afcf07d5af4a391104ba469f55f6304496933e2/Sources/WalletConnectNetworking/ProtocolMethod.swift#L4) into a `static var` however other parts of the codebase were depending on this being an instance `var`.

File: Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift (hunk 4 above)
Comment (jackpooleywc) [reply Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift:39]: ☝️ https://github.com/WalletConnect/WalletConnectSwiftV2/pull/1364#discussion_r1619024300

File: Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift (hunk 5 above)
Comment (jackpooleywc) [reply Sources/WalletConnectSign/Auth/Services/Wallet/AuthRequestSubscriber.swift:76]: Done: ✅ https://github.com/WalletConnect/WalletConnectSwiftV2/pull/1364/commits/e739136d1afff966c403aa4fe8243da95d35b73e

File: Sources/WalletConnectSign/Engine/Common/ApproveEngine.swift (hunk 6 above)
Comment (jackpooleywc) [reply Sources/WalletConnectSign/Engine/Common/ApproveEngine.swift:226]: ☝️ https://github.com/WalletConnect/WalletConnectSwiftV2/pull/1364#discussion_r1619024300

**Review by llbartekll — APPROVED** [review]

**Pull request merged**